src/kms/DeterministicTenantSecurityClient.ts

```diff
@@ -0,0 +1,116 @@
+        }
+        if (!apiKey) {
+            throw new Error("No value provided for TSP API Key");
+        }
+        this.tspDomain = tspDomain;
+        this.apiKey = apiKey;
+    }
+
+    /**
+     * Determine if the provided bytes are a deterministically encrypted field.
+     */
+    isCiphertext = (bytes: Buffer): boolean => Util.isDeterministicEncryptedField(bytes);
+
+    /**
+     * Deterministically encrypt the provided field using the tenant's current secret.
+     */
+    encryptField = (field: DeterministicPlaintextField, metadata: FieldMetadata): Promise<DeterministicEncryptedField> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, {[field.secretPath]: [field.derivationPath]}, metadata)
+            .flatMap((deriveResponse) => DetCrypto.encryptField(field, getDerivedKeys(deriveResponse, field.secretPath, field.derivationPath)))
+            .toPromise();
+
+    /**
+     * Deterministically encrypt a batch of new fields using the tenant's primary KMS. Supports partial failure and returns a list of fields that were successfully
+     * encrypted as well as a list of errors for fields that failed to be encrypted.
+     */
+    encryptFieldBatch = (fields: DeterministicPlaintextFieldCollection, metadata: FieldMetadata): Promise<BatchResult<DeterministicEncryptedField>> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, deterministicCollectionToPathMap(fields), metadata)
+            .flatMap((deriveResponse) => DetCrypto.batchEncryptField(fields, deriveResponse))
+            .toPromise();
+
+    /**
+     * Decrypt the provided deterministically encrypted field.
+     */
+    decryptField = (encryptedDoc: DeterministicEncryptedField, metadata: FieldMetadata): Promise<DeterministicPlaintextField> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, {[encryptedDoc.secretPath]: [encryptedDoc.derivationPath]}, metadata)
+            .flatMap((deriveResponse) =>
+                DetCrypto.decryptField(encryptedDoc, getDerivedKeys(deriveResponse, encryptedDoc.secretPath, encryptedDoc.derivationPath))
+            )
+            .toPromise();
+
+    /**
+     * Deterministically decrypt a batch of fields using the tenant's KMS that was used for encryption. Supports partial failure and will return both
+     * successfully decrypted fields as well as fields that failed to be decrypted.
+     */
+    decryptFieldBatch = (fields: DeterministicEncryptedFieldCollection, metadata: FieldMetadata): Promise<BatchResult<DeterministicPlaintextField>> =>
+        KmsApi.deriveKey(this.tspDomain, this.apiKey, deterministicCollectionToPathMap(fields), metadata)
+            .flatMap((deriveResponse) => DetCrypto.batchDecryptField(fields, deriveResponse))
+            .toPromise();
+
+    /**
+     * Decrypt the provided deterministically encrypted field and re-encrypt it with the current tenant secret.
+     * This should be called when rotating from one KMS configuration to another.
```

Comment:
I don't think this comment is correct. It's when they want to rotate from a tenant secret to another. Migration to different KMS configurations are handled automatically.

Reply:
Yep, you're right